src/python/pants/core/goals/clean.py

```diff
@@ -0,0 +1,158 @@
+# Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import annotations
+
+import os
+import tempfile
+from pathlib import Path
+from typing import Union
+
+import humanize
+
+from pants.base.build_environment import get_pants_cachedir
+from pants.core.util_rules.distdir import DistDir
+from pants.engine.console import Console
+from pants.engine.goal import Goal, GoalSubsystem
+from pants.engine.internals.options_parsing import _Options
+from pants.engine.rules import collect_rules, goal_rule
+from pants.option.global_options import LocalStoreOptions
+from pants.util.dirutil import rm_rf
+
+# N.B. We use a list so that named-caches can go last, as it is quite large and for dry-run
+# it's a nicer UX if it's last.
+_NAMED_CACHE_FLAGS_INFO = [
+    (
+        "process-execution-temp-dirs",
+        "Whether to keep the process execution temporary directories. These are normally cleaned up"
+        " automatically, however they are leaked if --no-process-execution-local-cleanup.",
+    ),
+    (
+        "lmdb-store",
+        "Whether to keep the local file store, which stores the results of subprocesses run by Pants.",
+    ),
+    (
+        "run-tracker",
+        "Whether to clean directories persisted after using `./pants run`.",
+    ),
+    (
+        "repl-temp-dirs",
+        "Whether to clean directories persisted after using `./pants repl`.",
+    ),
+    ("dist-dir", "Whether to clean the distributable files directory. "),
+    ("bootstrap-dir", "Whether to clean directories used by pants for bootstrapping itself."),
```

Comment:
Oh I should note too, this one is weird-ish because `pants` will be cleaning itself up. I thought about leaving the current pants' bootstrapped files but determined it'd just be more work.

Reply:
Aren't the versions of python pants runs on all with dicts that preserve insertion order now?

Reply:
(@kaos I think that comment belong in it's own thread, right?)